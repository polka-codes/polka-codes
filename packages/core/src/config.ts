import { z } from 'zod'

const providerModelSchema = z.object({
  provider: z.string().optional(),
  model: z.string().optional(),
  parameters: z.record(z.string(), z.any()).optional(),
})

const agentSchema = providerModelSchema.extend({
  initialContext: z
    .object({
      maxFileCount: z.number().int().positive().optional(),
      excludes: z.array(z.string()).optional(),
    })
    .optional(),
})

export const configSchema = z
  .object({
    agent: z.string().optional(),
    prices: z
      .record(
        z.string(), // provider
        z.record(
          z.string(), // model
          z.object({
            inputPrice: z.number().optional(),
            outputPrice: z.number().optional(),
            cacheWritesPrice: z.number().optional(),
            cacheReadsPrice: z.number().optional(),
          }),
        ),
      )
      .optional(),
    providers: z
      .record(
        z.string(),
        z.object({
          apiKey: z.string().optional(),
          defaultModel: z.string().optional(),
          defaultParameters: z.record(z.string(), z.any()).optional(),
        }),
      )
      .optional(),
    defaultProvider: z.string().optional(),
    defaultModel: z.string().optional(),
    defaultParameters: z.record(z.string(), z.any()).optional(),
    maxMessageCount: z.number().int().positive().optional(),
    budget: z.number().positive().optional(),
    scripts: z
      .record(
        z.string(),
        z.string().or(
          z.object({
            command: z.string(),
            description: z.string(),
          }),
        ),
      )
      .optional(),
    agents: z.record(z.string(), agentSchema).optional(),
    commands: z
      .object({
        default: providerModelSchema.optional(),
      })
      .catchall(providerModelSchema)
      .optional(),
    rules: z.array(z.string()).optional().or(z.string()).optional(),
    excludeFiles: z.array(z.string()).optional(),
  })
  .strict()

export type Config = z.infer<typeof configSchema>
