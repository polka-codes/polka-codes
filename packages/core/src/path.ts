/**
 * Returns the directory portion of a path string.
 * Strips trailing slashes, then takes everything up to the last slash.
 */
export function dirname(path: string): string {
  if (path.length === 0) return '.'
  const isRooted = path[0] === '/'
  // Trim trailing slashes (but leave root "/")
  let end = path.length - 1
  while (end > 0 && path[end] === '/') end--
  const idx = path.lastIndexOf('/', end)
  if (idx < 0) {
    return isRooted ? '/' : '.'
  }
  if (isRooted && idx === 0) {
    return '/'
  }
  return path.slice(0, idx)
}

/**
 * Normalizes a path by resolving "." and ".." segments.
 */
function normalize(path: string): string {
  const isAbsolute = path.startsWith('/')
  const segments = path.split('/').filter(Boolean)
  const stack: string[] = []
  for (const seg of segments) {
    if (seg === '.') continue
    if (seg === '..') {
      if (stack.length && stack[stack.length - 1] !== '..') {
        stack.pop()
      } else if (!isAbsolute) {
        stack.push('..')
      }
    } else {
      stack.push(seg)
    }
  }
  let result = stack.join('/')
  if (!result && !isAbsolute) return '.'
  if (result && path.endsWith('/')) result += '/'
  return (isAbsolute ? '/' : '') + result
}

/**
 * Joins all given path segments, then normalizes the result.
 */
export function join(...parts: string[]): string {
  if (parts.length === 0) return '.'
  let combined = ''
  for (const p of parts) {
    if (typeof p !== 'string') {
      throw new TypeError('Arguments to join must be strings')
    }
    if (p) {
      combined = combined ? `${combined}/${p}` : p
    }
  }
  return normalize(combined)
}
