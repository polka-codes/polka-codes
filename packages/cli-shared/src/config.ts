// Generated by polka.codes

import { existsSync, readFileSync } from 'node:fs'
import { homedir } from 'node:os'
import { join } from 'node:path'
import { type Config, configSchema } from '@polka-codes/core'
import { merge } from 'lodash-es'
import { parse } from 'yaml'
import { ZodError } from 'zod'

export { configSchema, type Config }

export function getGlobalConfigPath(home = homedir()): string {
  return join(home, '.config', 'polkacodes', 'config.yml')
}

export function loadConfigAtPath(path: string): Config | undefined {
  try {
    return readConfig(path)
  } catch (_error) {
    return undefined
  }
}

export const localConfigFileName = '.polkacodes.yml'

const mergeArray = (a: string[] | undefined, b: string[] | undefined): string[] | undefined => {
  if (!a && !b) {
    return undefined
  }
  if (!a) {
    return b
  }
  if (!b) {
    return a
  }
  return [...a, ...b]
}

export function mergeConfigs(configs: Config[]): Config {
  if (configs.length === 0) {
    return {}
  }

  const mergedConfig = configs.reduce((acc, config) => {
    const merged = merge({}, acc, config)

    // Special handling for rules array
    let accRules = acc.rules ?? []
    if (typeof accRules === 'string') {
      accRules = [accRules]
    }
    let configRules = config.rules ?? []
    if (typeof configRules === 'string') {
      configRules = [configRules]
    }
    merged.rules = mergeArray(accRules, configRules)

    // Special handling for excludeFiles array
    merged.excludeFiles = mergeArray(acc.excludeFiles, config.excludeFiles)

    return merged
  })

  return mergedConfig
}

export function loadConfig(paths?: string | string[], cwd: string = process.cwd(), home = homedir()): Config | undefined {
  const configs: Config[] = []

  // Load global config if exists
  const globalConfigPath = getGlobalConfigPath(home)
  if (existsSync(globalConfigPath)) {
    try {
      const globalConfig = readConfig(globalConfigPath)
      configs.push(globalConfig)
    } catch (error) {
      console.warn(`Error loading global config file: ${globalConfigPath}\n${error}`)
    }
  }

  // Load project configs
  if (paths && paths.length > 0) {
    const configPaths = Array.isArray(paths) ? paths : [paths]
    for (const path of configPaths) {
      try {
        const config = readConfig(path)
        configs.push(config)
      } catch (error) {
        console.error(`Error loading config file: ${path}\n${error}`)
        throw error
      }
    }
  } else {
    const configPath = join(cwd, localConfigFileName)
    try {
      const projectConfig = readConfig(configPath)
      configs.push(projectConfig)
    } catch (error) {
      if (error instanceof ZodError) {
        console.error(`Error in config file: ${configPath}\n${error}`)
        throw error
      }
    }
  }

  return configs.length > 0 ? mergeConfigs(configs) : undefined
}

export const readConfig = (path: string): Config => {
  const file = readFileSync(path, 'utf8')
  const config = parse(file)
  return configSchema.parse(config)
}

export const readLocalConfig = (path?: string): Config | undefined => {
  try {
    return readConfig(path ?? localConfigFileName)
  } catch (_error) {
    return undefined
  }
}
