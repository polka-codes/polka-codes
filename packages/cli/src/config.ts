import { existsSync, readFileSync } from 'node:fs'
import { homedir } from 'node:os'
import { join } from 'node:path'
import { merge } from 'lodash'
import { parse } from 'yaml'
import { ZodError, z } from 'zod'

import { type AgentNameType, allAgents } from '@polka-codes/core'

const agentNameValues = allAgents.map((agent) => agent.name)

// TODO: figure out if there is a better way to do this
const agentNames = z.enum<AgentNameType, [AgentNameType, ...AgentNameType[]]>([agentNameValues[0], ...agentNameValues.slice(1)])
const agentNameOrDefault = z.union([agentNames, z.literal('default')])

const providerModelSchema = z.object({
  provider: z.string().optional(),
  model: z.string().optional(),
})

const agentSchema = providerModelSchema.extend({
  initialContext: z
    .object({
      maxFileCount: z.number().int().positive().optional(),
      excludes: z.array(z.string()).optional(),
    })
    .optional(),
})

export const configSchema = z
  .object({
    providers: z
      .record(
        z.string(),
        z.object({
          apiKey: z.string().optional(),
          defaultModel: z.string().optional(),
        }),
      )
      .optional(),
    defaultProvider: z.string().optional(),
    defaultModel: z.string().optional(),
    maxMessageCount: z.number().int().positive().optional(),
    budget: z.number().positive().optional(),
    scripts: z
      .record(
        z.string(),
        z.string().or(
          z.object({
            command: z.string(),
            description: z.string(),
          }),
        ),
      )
      .optional(),
    agents: z.record(agentNameOrDefault, agentSchema).optional(),
    commands: z
      .object({
        default: providerModelSchema.optional(),
      })
      .catchall(providerModelSchema)
      .optional(),
    rules: z.array(z.string()).optional().or(z.string()).optional(),
    excludeFiles: z.array(z.string()).optional(),
  })
  .strict()

export type Config = z.infer<typeof configSchema>

export function getGlobalConfigPath(home = homedir()): string {
  return join(home, '.config', 'polkacodes', 'config.yml')
}

export function loadConfigAtPath(path: string): Config | undefined {
  try {
    return readConfig(path)
  } catch (error) {
    return undefined
  }
}

export const localConfigFileName = '.polkacodes.yml'

const mergeArray = (a: string[] | undefined, b: string[] | undefined): string[] | undefined => {
  if (!a && !b) {
    return undefined
  }
  if (!a) {
    return b
  }
  if (!b) {
    return a
  }
  return [...a, ...b]
}

export function loadConfig(path?: string, cwd: string = process.cwd(), home = homedir()): Config | undefined {
  // Load global config if exists
  const globalConfigPath = getGlobalConfigPath(home)
  let globalConfig: Config | undefined
  if (existsSync(globalConfigPath)) {
    try {
      globalConfig = readConfig(globalConfigPath)
    } catch (error) {
      console.warn(`Error loading global config file: ${globalConfigPath}\n${error}`)
    }
  }

  // Load project config
  let projectConfig: Config | undefined
  if (path) {
    try {
      projectConfig = readConfig(path)
    } catch (error) {
      console.error(`Error loading config file: ${path}\n${error}`)
      throw error
    }
  } else {
    const configPath = join(cwd, localConfigFileName)
    try {
      projectConfig = readConfig(configPath)
    } catch (error) {
      if (error instanceof ZodError) {
        console.error(`Error in config file: ${path}\n${error}`)
        throw error
      }
    }
  }

  // Merge configs with project config taking precedence
  if (globalConfig && projectConfig) {
    const mergedConfig = merge({}, globalConfig, projectConfig)

    // Merge rules array
    let projectRules = projectConfig.rules ?? []
    if (typeof projectRules === 'string') {
      projectRules = [projectRules]
    }
    let globalRules = globalConfig.rules ?? []
    if (typeof globalRules === 'string') {
      globalRules = [globalRules]
    }
    mergedConfig.rules = mergeArray(globalRules, projectRules)

    // Merge exclude files
    mergedConfig.excludeFiles = mergeArray(globalConfig.excludeFiles, projectConfig.excludeFiles)

    return mergedConfig
  }
  return projectConfig || globalConfig
}

export const readConfig = (path: string): Config => {
  const file = readFileSync(path, 'utf8')
  const config = parse(file)
  return configSchema.parse(config)
}

export const readLocalConfig = (path?: string): Config | undefined => {
  try {
    return readConfig(path ?? localConfigFileName)
  } catch (error) {
    return undefined
  }
}
