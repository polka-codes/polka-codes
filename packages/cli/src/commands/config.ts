import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs'
import { dirname } from 'node:path'
import { input, password, select } from '@inquirer/prompts'
import { stringify } from 'yaml'

import { AiServiceProvider, anthropicDefaultModelId, anthropicModels, deepSeekDefaultModelId } from '@polka-codes/core'

import { set } from 'lodash'
import { getGlobalConfigPath, loadConfigAtPath, localConfigFileName } from '../config'

const fetchOllamaModels = async () => {
  try {
    const resp = await fetch('http://localhost:11434/api/tags')
    const data = await resp.json()
    return data.models.map((model: any) => model.name) as string[]
  } catch (error) {
    console.log('Unable to fetch Ollama models')
    return []
  }
}

type ProviderConfig = {
  provider: string
  model: string
  apiKey?: string
}

export async function configPrompt(existingConfig?: Partial<ProviderConfig>): Promise<ProviderConfig> {
  // select AI provider
  const provider = await select({
    message: 'Choose AI Provider:',
    choices: Object.entries(AiServiceProvider).map(([key, value]) => ({ name: key, value })),
    default: existingConfig?.provider,
  })

  let model = existingConfig?.model

  switch (provider) {
    case AiServiceProvider.Anthropic:
      model = await select({
        message: 'Choose Model ID:',
        choices: Object.keys(anthropicModels).map((key) => ({ name: key, value: key })),
        default: existingConfig?.model ?? anthropicDefaultModelId,
      })
      break
    case AiServiceProvider.Ollama:
      {
        const models = await fetchOllamaModels()
        if (models && models.length > 0) {
          model = await select({
            message: 'Choose Model ID:',
            choices: models.map((model) => ({ name: model, value: model })),
            default: existingConfig?.model,
          })
        } else {
          model = await input({ message: 'Enter Model ID:' })
        }
      }
      // fetch model list from Ollama API

      break
    case AiServiceProvider.DeepSeek:
      model = deepSeekDefaultModelId
      break
  }

  let apiKey: string | undefined
  if (provider !== AiServiceProvider.Ollama) {
    apiKey = await password({ message: 'Enter API Key:', mask: '*' })
  }

  return { provider, model, apiKey }
}

async function printConfig(configPath: string) {
  const config = existsSync(configPath) ? loadConfigAtPath(configPath) : undefined
  if (!config) {
    console.log('No config found')
    return
  }
  console.log(`Config file path: ${configPath}`)
  if (config.providers) {
    for (const key of Object.keys(config.providers)) {
      if (config.providers[key as AiServiceProvider].apiKey) {
        config.providers[key as AiServiceProvider].apiKey = '<redacted>'
      }
    }
  }

  console.log(stringify(config))
}

export async function configCommand(options: { global?: boolean; print?: boolean }) {
  const globalConfigPath = getGlobalConfigPath()
  let configPath = options.global ? globalConfigPath : localConfigFileName

  if (options.print) {
    await printConfig(configPath)
    return
  }

  const existingConfig = existsSync(configPath) ? loadConfigAtPath(configPath) : undefined

  if (!existingConfig && !options.global) {
    const isGlobal = await select({
      message: 'No config file found. Do you want to create one?',
      choices: [
        {
          name: `Create a global config at ${globalConfigPath}`,
          value: true,
        },
        {
          name: `Create a local config at ${configPath}`,
          value: false,
        },
      ],
    })
    if (isGlobal) {
      configPath = globalConfigPath
    } else {
      configPath = localConfigFileName
    }
  }
  console.log(`Config file path: ${configPath}`)

  let { provider, model, apiKey } = await configPrompt({ provider: existingConfig?.defaultProvider, model: existingConfig?.defaultModel })

  if (apiKey && configPath === localConfigFileName) {
    const option = await select({
      message: 'It is not recommended to store API keys in the local config file. How would you like to proceed?',
      choices: [
        {
          name: 'Save API key in the local config file',
          value: 1,
        },
        {
          name: 'Save API key in the global config file',
          value: 2,
        },
        {
          name: 'Save API key to .env file',
          value: 3,
        },
      ],
    })

    switch (option) {
      case 1:
        // do nothing
        break
      case 2: {
        const globalConfig = loadConfigAtPath(globalConfigPath) ?? {}
        set(globalConfig, ['providers', provider, 'apiKey'], apiKey)
        mkdirSync(dirname(globalConfigPath), { recursive: true })
        writeFileSync(globalConfigPath, stringify(globalConfig))
        console.log(`API key saved to global config file: ${globalConfigPath}`)
        apiKey = undefined
        break
      }
      case 3: {
        let envFileContent: string
        if (existsSync('.env')) {
          envFileContent = readFileSync('.env', 'utf8')
          envFileContent += `\n${provider.toUpperCase()}_API_KEY=${apiKey}`
        } else {
          envFileContent = `${provider.toUpperCase()}_API_KEY=${apiKey}`
        }
        writeFileSync('.env', envFileContent)
        console.log('API key saved to .env file')
        apiKey = undefined
        break
      }
    }
  }

  const newConfig = {
    ...existingConfig,
    defaultProvider: provider,
    defaultModel: model,
  }
  if (apiKey) {
    set(newConfig, ['providers', provider, 'apiKey'], apiKey)
  }
  mkdirSync(dirname(configPath), { recursive: true })
  writeFileSync(configPath, stringify(newConfig))
  console.log(`Config file saved at: ${configPath}`)
  return
}
