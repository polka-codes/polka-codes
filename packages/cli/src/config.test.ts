import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, test } from 'bun:test'
import { mkdirSync, mkdtempSync, rmSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { loadConfig } from './config'

describe('config', () => {
  let testDir: string
  let testSubDir: string
  let testHomeDir: string

  beforeAll(() => {
    testDir = mkdtempSync(join(tmpdir(), 'polkacodes-test-'))
  })

  beforeEach(() => {
    testSubDir = mkdtempSync(join(testDir, `sub-test-${Math.random()}`))
    testHomeDir = mkdtempSync(join(testDir, `home-test-${Math.random()}`))
    mkdirSync(join(testHomeDir, '.config', 'polkacodes'), { recursive: true })
  })

  afterEach(() => {
    rmSync(testSubDir, { recursive: true, force: true })
    rmSync(testHomeDir, { recursive: true, force: true })
  })

  afterAll(() => {
    rmSync(testDir, { recursive: true, force: true })
  })

  test('loads config from specified path', () => {
    const configPath = join(testSubDir, 'test-config.yml')
    writeFileSync(
      configPath,
      `
defaultProvider: anthropic
defaultModel: claude-3-opus
    `,
    )

    const config = loadConfig(configPath, testSubDir, testHomeDir)
    expect(config).toEqual({
      defaultProvider: 'anthropic',
      defaultModel: 'claude-3-opus',
    })
  })

  test('loads config from default paths', () => {
    const configPath = join(testSubDir, '.polkacodes.yml')
    writeFileSync(
      configPath,
      `
defaultProvider: deepseek
defaultModel: deepseek-chat
    `,
    )

    const config = loadConfig(undefined, testSubDir, testHomeDir)
    expect(config).toEqual({
      defaultProvider: 'deepseek',
      defaultModel: 'deepseek-chat',
    })

    // Clean up
    writeFileSync(configPath, '')
  })

  test('validates config schema', () => {
    const configPath = join(testSubDir, 'invalid-config.yml')
    writeFileSync(
      configPath,
      `
invalidKey: value
    `,
    )

    expect(() => loadConfig(configPath, testSubDir, testHomeDir)).toThrow()
  })

  test('handles commands configuration', () => {
    const configPath = join(testSubDir, 'commands-config.yml')
    writeFileSync(
      configPath,
      `
scripts:
  test: echo "test"
  complex:
    command: echo "complex"
    description: A complex command
    `,
    )

    const config = loadConfig(configPath, testSubDir)
    expect(config?.scripts).toEqual({
      test: 'echo "test"',
      complex: {
        command: 'echo "complex"',
        description: 'A complex command',
      },
    })
  })

  test('handles rules configuration', () => {
    const configPath = join(testSubDir, 'rules-config.yml')
    writeFileSync(
      configPath,
      `
rules:
  - rule1
  - rule2
    `,
    )

    const config = loadConfig(configPath, testSubDir)
    expect(config?.rules).toEqual(['rule1', 'rule2'])
  })

  test('merges global and local config with local precedence', () => {
    const globalConfigPath = join(testHomeDir, '.config', 'polkacodes', 'config.yml')
    const localConfigPath = join(testSubDir, '.polkacodes.yml')

    writeFileSync(
      globalConfigPath,
      `
defaultProvider: anthropic
defaultModel: claude-3-opus
scripts:
  test: echo "global"
  complex:
    command: echo "global-complex"
    description: Global complex command
rules:
  - global-rule
    `,
    )

    writeFileSync(
      localConfigPath,
      `
providers:
  anthropic:
    apiKey: local-key
scripts:
  test: echo "local"
rules:
  - local-rule
    `,
    )

    const config = loadConfig(localConfigPath, testSubDir, testHomeDir)
    expect(config).toEqual({
      defaultProvider: 'anthropic',
      defaultModel: 'claude-3-opus',
      excludeFiles: undefined,
      providers: {
        anthropic: {
          apiKey: 'local-key',
        },
      },
      scripts: {
        test: 'echo "local"',
        complex: {
          command: 'echo "global-complex"',
          description: 'Global complex command',
        },
      },
      rules: ['global-rule', 'local-rule'],
    })
  })

  test('concatenates arrays', () => {
    const globalConfigPath = join(testHomeDir, '.config', 'polkacodes', 'config.yml')
    const localConfigPath = join(testSubDir, '.polkacodes.yml')

    writeFileSync(
      globalConfigPath,
      `
rules:
  - global-rule-1
  - global-rule-2
    `,
    )

    writeFileSync(
      localConfigPath,
      `
rules: local-rule
    `,
    )

    const config = loadConfig(localConfigPath, testSubDir, testHomeDir)
    expect(config?.rules).toEqual(['global-rule-1', 'global-rule-2', 'local-rule'])
  })
})
