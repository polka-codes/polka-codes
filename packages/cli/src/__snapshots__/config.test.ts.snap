// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`config merges global and local config with local precedence 1`] = `
{
  "agents": {
    "architect": {
      "provider": "deepseek",
    },
    "coder": {
      "model": "claude-3-haiku",
    },
    "default": {
      "model": "claude-3-opus",
      "provider": "anthropic",
    },
  },
  "commands": {
    "default": {
      "model": "deepseek-coder-instruct",
      "provider": "deepseek",
    },
  },
  "defaultModel": "claude-3-opus",
  "defaultProvider": "anthropic",
  "excludeFiles": undefined,
  "hooks": {
    "agents": {
      "architect": {
        "beforeCompletion": "echo 'local architect hook'",
      },
      "coder": {
        "beforeCompletion": "echo 'local coder hook'",
      },
    },
  },
  "providers": {
    "anthropic": {
      "apiKey": "local-key",
    },
  },
  "rules": [
    "global-rule",
    "local-rule",
  ],
  "scripts": {
    "complex": {
      "command": "echo "global-complex"",
      "description": "Global complex command",
    },
    "test": "echo "local"",
  },
}
`;

exports[`config parses project .polkacodes.yml successfully 1`] = `
{
  "agents": {
    "default": {
      "initialContext": {
        "excludes": [
          "__snapshots__",
          ".husky",
        ],
      },
    },
  },
  "excludeFiles": [
    ".env",
    "bun.lock",
    "packages/cli/cli.mjs",
  ],
  "hooks": {
    "agents": {
      "coder": {
        "beforeCompletion": "bun fix; bun typecheck && bun test -u",
      },
    },
  },
  "rules": 
"Use \`bun\` as the package manager
When adding new dependencies, cd to the package directory and run \`bun add <dependency>\`
Use \`bun:test\` to write tests. DO NOT use jest or mocha or vi
DO NOT use mock in tests. Use spyOn imported bun 'bun:test' if needed
When writing unit tests, use \`.toMatchSnapshot()\` if suitable
When creating a new file, try to check if there are other files with similar content, and if so, read those files and use them as a reference
    
"
,
  "scripts": {
    "test": {
      "command": "bun test",
      "description": "Run tests. Pass file path to run a specific test file.",
    },
    "typecheck": {
      "command": "bun typecheck",
      "description": 
"Check the code for type errors.
Run this command after making changes to the code.
"
,
    },
  },
}
`;

exports[`config parses example.polkacodes.yml successfully 1`] = `
{
  "agents": {
    "architect": {
      "model": "deepseek-reasoner",
      "provider": "deepseek",
    },
    "coder": {
      "model": "claude-3-5-sonnet-20241022",
      "provider": "anthropic",
    },
    "default": {
      "initialContext": {
        "excludes": [
          "__snapshots__",
          "*.lock",
        ],
        "maxFileCount": 200,
      },
      "model": "deepseek-chat",
      "provider": "deepseek",
    },
  },
  "commands": {
    "commit": {
      "model": "maryasov/qwen2.5-coder-cline:7b",
      "provider": "ollama",
    },
    "default": {
      "model": "deepseek-chat",
      "provider": "deepseek",
    },
    "pr": {
      "model": "maryasov/qwen2.5-coder-cline:7b",
      "provider": "ollama",
    },
  },
  "defaultModel": "deepseek-chat",
  "defaultProvider": "deepseek",
  "excludeFiles": [
    "node_modules/**",
    "dist/**",
    "*.log",
    ".env",
    "coverage/**",
  ],
  "hooks": {
    "agents": {
      "architect": {
        "beforeCompletion": "bun test && bun typecheck",
      },
      "coder": {
        "beforeCompletion": "bun typecheck",
      },
      "default": {
        "beforeCompletion": "bun test",
      },
    },
  },
  "maxIterations": 50,
  "providers": {
    "anthropic": {
      "apiKey": "sk-ant-...",
      "defaultModel": "claude-3-5-sonnet-20241022",
    },
    "deepseek": {
      "apiKey": "sk-...",
      "defaultModel": "deepseek-chat",
    },
    "ollama": {
      "defaultModel": "maryasov/qwen2.5-coder-cline:7b",
    },
    "openrouter": {
      "apiKey": "sk-...",
      "defaultModel": "openai/gpt-4o",
    },
  },
  "rules": [
    "Use \`bun\` as the package manager",
    "Write tests using bun:test",
    "Follow the project's coding style",
    "Add JSDoc comments for public APIs",
  ],
  "scripts": {
    "build": {
      "command": "bun build",
      "description": "Build the project",
    },
    "test": "bun test",
    "typecheck": {
      "command": "bun typecheck",
      "description": "Run TypeScript type checking",
    },
  },
}
`;
